ui/dashboards/src/context/DashboardProvider/panel-group-slice.ts

```diff
@@ -76,18 +89,36 @@ export interface PanelGroupItemId {
 export function createPanelGroupSlice(
   layouts: LayoutDefinition[]
 ): StateCreator<PanelGroupSlice, Middleware, [], PanelGroupSlice> {
-  // Convert the initial layouts from the JSON to panel groups and keep track of the order
+  // Convert the initial layouts from the JSON
   const panelGroups: PanelGroupSlice['panelGroups'] = {};
   const panelGroupIdOrder: PanelGroupSlice['panelGroupOrder'] = [];
   for (const layout of layouts) {
-    const id = generateId();
-    panelGroups[id] = {
-      id,
-      items: layout.spec.items,
+    const itemLayouts: PanelGroupDefinition['itemLayouts'] = [];
+    const itemPanelKeys: PanelGroupDefinition['itemPanelKeys'] = {};
+
+    // Split layout information from panel keys to make it easier to update just layouts on move/resize of panels
+    for (const item of layout.spec.items) {
+      const panelGroupLayoutId = generateId().toString();
```

Comment:
i am thinking if we could somehow generate/enforce unique `panelRef`, then we wouldn't have to keep track of unique id and panelRef

Reply:
Yeah, we're still dealing with the whole "panels can be referenced multiple times" thing, so a `panelRef` isn't guaranteed to be unique. 🙃